controllers/secret_updater.go

```diff
@@ -123,65 +125,86 @@ func (p *updateProcess) stop(ctx context.Context) error {
 }
 
 func (p *updateProcess) run(ctx context.Context) {
-	p.retryCountMetrics.Set(0)
-	p.log.Info("wait for an empty secret to be created by reconcile")
+	p.log.Info("start a secret updater process")
+	waitTime := 1 * time.Second
 	for {
-		_, err := p.getSecret(ctx)
-		if err == nil {
-			break
-		}
 		select {
 		case <-ctx.Done():
 			p.log.Info("stop a secret updater process")
 			return
-		case <-time.After(1 * time.Second):
+		case <-time.After(waitTime):
 		}
-	}
 
-	p.log.Info("start a secret updater process")
-	updateTime := 1 * time.Second
-	for {
-		select {
-		case <-ctx.Done():
-			p.log.Info("stop a secret updater process")
-			return
-		case <-time.After(updateTime):
-			err := p.updateSecret(ctx)
-			if err != nil {
-				p.log.Error(err, "failed to update secret, retry after 5 minutes")
-				p.retryCountMetrics.Inc()
-				updateTime = 5 * time.Minute
-				continue
-			}
+		s, err := p.getSecret(ctx)
+		if apierrors.IsNotFound(err) {
+			p.log.Error(err, "secret is not found")
+			waitTime = 1 * time.Second
+			continue
+		} else if err != nil {
+			p.log.Error(err, "failed to get secret, retry after 5 minutes")
+			p.retryCountMetrics.Inc()
+			waitTime = 5 * time.Minute
+			continue
 		}
-		expiresAt, err := p.getExpiresAt(ctx)
+
+		if need, updateTime := p.needUpdate(s); !need {
+			p.log.Info("wait until next update time", "updateTime", updateTime.Format(time.RFC3339))
+			waitTime = time.Until(updateTime)
+			continue
+		}
+
+		expiresAt, err := p.updateSecret(ctx, s)
 		if err != nil {
-			p.log.Error(err, "failed to get expires-at, retry after 5 minutes")
+			p.log.Error(err, "failed to update secret, retry after 5 minutes")
 			p.retryCountMetrics.Inc()
-			updateTime = 5 * time.Minute
+			waitTime = 5 * time.Minute
```

Comment:
I think 5 min waitTime is long, because after 5 minutes, the token will expire.
How about 1 minute?